sql: Add support for numeric JSON scalar casts

chrisseto opened the pull request:
* Fixes #41333
* Add support for queries such as
  - `SELECT '1'::jsonb::int`
  - `SELECT '1'::jsonb::float`
  - `SELECT '3.14'::jsonb::decimal`

Release note (sql change): Casting JSON numeric scalars to numeric
types now works as expected.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41367)
<!-- Reviewable:end -->

justinj reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41367#-:-Lqc8nKo9tsT8Ncy_EUO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chrisseto)

___
*[pkg/sql/sem/tree/eval.go, line 3219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41367#-Lqc4mz2Cm8NUPxn5AVa:-Lqc4mz3-xa_mmPggTnq:b-ordya4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf9ee38d09c5cc58efb96784c60768e003119cd/pkg/sql/sem/tree/eval.go#L3219)):*
> ```Go
> 			if v.Type() == json.NumberJSONType {
> 				// err is ignored as a more appropriate error
> 				// will be generated later
> ```

nit: periods on comments (same below)

also, if we're always(?) just going to ignore this error, can we just not construct it in the first place and return a bool indicating whether we were able to successfully do the conversion?
___
*[pkg/sql/sem/tree/testdata/eval/cast, line 1034 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41367#-Lqc5ozGB-lA3a1vJS0f:-Lqc5ozGB-lA3a1vJS0g:b-cef74q) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf9ee38d09c5cc58efb96784c60768e003119cd/pkg/sql/sem/tree/testdata/eval/cast#L1034)):*
> ```
> '3.14'::jsonb::string
> ----
> '3.14'
> ```

nit: this could include some invalid cases, and also some that invoke truncation(?) like `'1.5'::json::int` (should compare to postgres to determine the proper behaviour).
___
*[pkg/util/json/json.go, line 975 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41367#-Lqc62fs7ySZ67jUQt1_:-Lqc62fs7ySZ67jUQt1a:b-f3fb32) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcf9ee38d09c5cc58efb96784c60768e003119cd/pkg/util/json/json.go#L975)):*
> ```Go
> }
> 
> // ToDecimal returns a apd.Decimal given a JSON value
> ```

nit: "...an apd.Decimal..." and period at the end of comment


<!-- Sent from Reviewable.io -->

jordanlewis commented:
Hey @chrisseto, any interest in finishing this fella up?

chrisseto commented:
@jordanlewis Hey! Sorry for the late response but yes I am. I ran into some issues with decimals IIRC.
Would you mind if I stole a bit of your time on a Friday?

jordanlewis commented:
Let's do it!

nvanbenschoten commented:
This PR looks great! I've always found the ergonomics of JSON to be a little off, and I think a lot of it has to do with the absence of these casts. @chrisseto any plans to polish this off and get it merged?

chrisseto commented:
Thanks for the ping @nvanbenschoten. I'll see if I can get it cleaned up this week!

chrisseto committed:
sql: Add support for numeric JSON scalar casts

Previously, casting from a JSON numeric value to numeric values was
unspported.
This commit adds support for this feature allowing queries such as:
`SELECT '1'::jsonb::int`
`SELECT '1'::jsonb::float`
`SELECT '3.14'::jsonb::decimal`
to work as expected.

Fixes #41333

Release note (sql change): Casting JSON numeric scalars to numeric
                           types now works as expected.

jordanlewis commented:
cc @rafiss

rafiss reviewed (approved):
this lgtm! thanks @chrisseto! looks it does need a rebase though

rafiss commented:
awesome!

bors r+

chrisseto commented:
TFRT @rafiss !!
